File: java/src/main/java/org/libpetri/debug/DebugResponse.java
package org.libpetri.debug;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Responses sent from server to debug UI client via WebSocket.
 *
 * <p>This sealed hierarchy represents all debug protocol responses.
 * Jackson polymorphic serialization includes a "type" field to identify
 * the response type.
 *
 * <h2>Protocol Example</h2>
 * <pre>{@code
 * // Session list
 * {"type": "sessionList", "sessions": [...]}
 *
 * // Subscribed to session
 * {"type": "subscribed", "sessionId": "abc-123", "dotDiagram": "...", ...}
 *
 * // Event received
 * {"type": "event", "sessionId": "abc-123", "event": {...}}
 * }</pre>
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(value = DebugResponse.SessionList.class, name = "sessionList"),
    @JsonSubTypes.Type(value = DebugResponse.Subscribed.class, name = "subscribed"),
    @JsonSubTypes.Type(value = DebugResponse.Unsubscribed.class, name = "unsubscribed"),
    @JsonSubTypes.Type(value = DebugResponse.Event.class, name = "event"),
    @JsonSubTypes.Type(value = DebugResponse.EventBatch.class, name = "eventBatch"),
    @JsonSubTypes.Type(value = DebugResponse.MarkingSnapshot.class, name = "markingSnapshot"),
    @JsonSubTypes.Type(value = DebugResponse.PlaybackStateChanged.class, name = "playbackStateChanged"),
    @JsonSubTypes.Type(value = DebugResponse.FilterApplied.class, name = "filterApplied"),
    @JsonSubTypes.Type(value = DebugResponse.BreakpointHit.class, name = "breakpointHit"),
    @JsonSubTypes.Type(value = DebugResponse.BreakpointList.class, name = "breakpointList"),
    @JsonSubTypes.Type(value = DebugResponse.BreakpointSet.class, name = "breakpointSet"),
    @JsonSubTypes.Type(value = DebugResponse.BreakpointCleared.class, name = "breakpointCleared"),
    @JsonSubTypes.Type(value = DebugResponse.Error.class, name = "error"),
    @JsonSubTypes.Type(value = DebugResponse.ArchiveList.class, name = "archiveList"),
    @JsonSubTypes.Type(value = DebugResponse.ArchiveImported.class, name = "archiveImported")
})
public sealed interface DebugResponse {

    /**
     * List of available debug sessions.
     *
     * @param sessions list of session summaries
     */
    record SessionList(List<SessionSummary> sessions) implements DebugResponse {}

    /**
     * Confirmation of subscription with initial state.
     *
     * <p>The {@code subnetInstances} field carries the post-composition
     * subnet instance descriptors per {@code spec/11-modular-composition.md}
     * <b>MOD-041</b>. For nets that were not produced via composition (no
     * {@code "/"} in any node name) the list is empty.
     *
     * @param sessionId the subscribed session
     * @param netName the Petri net name
     * @param dotDiagram DOT (Graphviz) diagram of the net
     * @param structure explicit net topology with authoritative name-to-graphId mappings
     * @param currentMarking current token distribution
     * @param enabledTransitions currently enabled transitions
     * @param inFlightTransitions currently executing transitions (started but not yet completed/failed/timed out)
     * @param eventCount total events in session
     * @param mode subscription mode (live or replay)
     * @param subnetInstances composed subnet instance descriptors (empty for flat nets); per <b>MOD-041</b>
     */
    record Subscribed(
        String sessionId,
        String netName,
        String dotDiagram,
        NetStructure structure,
        Map<String, List<TokenInfo>> currentMarking,
        List<String> enabledTransitions,
        List<String> inFlightTransitions,
        long eventCount,
        String mode,
        List<SubnetInstanceInfo> subnetInstances
    ) implements DebugResponse {
        /**
         * Backwards-compatible constructor used by older call sites that
         * pre-date the {@code subnetInstances} field. Always passes an empty
         * list so wire shape stays stable for non-composed nets.
         */
        public Subscribed(
            String sessionId,
            String netName,
            String dotDiagram,
            NetStructure structure,
            Map<String, List<TokenInfo>> currentMarking,
            List<String> enabledTransitions,
            List<String> inFlightTransitions,
            long eventCount,
            String mode
        ) {
            this(sessionId, netName, dotDiagram, structure, currentMarking,
                enabledTransitions, inFlightTransitions, eventCount, mode, List.of());
        }
    }

    /**
     * Confirmation of unsubscription.
     *
     * @param sessionId the unsubscribed session
     */
    record Unsubscribed(String sessionId) implements DebugResponse {}

    /**
     * Single event notification.
     *
     * @param sessionId the session this event belongs to
     * @param index event index in the session
     * @param event the event data
     */
    record Event(
        String sessionId,
        long index,
        NetEventInfo event
    ) implements DebugResponse {}

    /**
     * Batch of events (for replay catch-up).
     *
     * @param sessionId the session these events belong to
     * @param startIndex index of first event in batch
     * @param events list of events
     * @param hasMore true if more events available
     */
    record EventBatch(
        String sessionId,
        long startIndex,
        List<NetEventInfo> events,
        boolean hasMore
    ) implements DebugResponse {}

    /**
     * Complete marking snapshot.
     *
     * @param sessionId the session this marking belongs to
     * @param marking map of place names to tokens
     * @param enabledTransitions currently enabled transitions
     * @param inFlightTransitions currently firing transitions
     */
    record MarkingSnapshot(
        String sessionId,
        Map<String, List<TokenInfo>> marking,
        List<String> enabledTransitions,
        List<String> inFlightTransitions
    ) implements DebugResponse {}

    /**
     * Playback state changed (pause/resume/speed).
     *
     * @param sessionId the session
     * @param paused whether playback is paused
     * @param speed current playback speed
     * @param currentIndex current event index
     */
    record PlaybackStateChanged(
        String sessionId,
        boolean paused,
        double speed,
        long currentIndex
    ) implements DebugResponse {}

    /**
     * Filter applied confirmation.
     *
     * @param sessionId the session the filter applies to
     * @param filter the applied filter configuration
     */
    record FilterApplied(
        String sessionId,
        DebugCommand.EventFilter filter
    ) implements DebugResponse {}

    /**
     * Breakpoint was hit - execution paused.
     *
     * @param sessionId the session where the breakpoint was hit
     * @param breakpointId the ID of the breakpoint that was hit
     * @param event the event that triggered the breakpoint
     * @param eventIndex the index of the event
     */
    record BreakpointHit(
        String sessionId,
        String breakpointId,
        NetEventInfo event,
        long eventIndex
    ) implements DebugResponse {}

    /**
     * List of breakpoints for a session.
     *
     * @param sessionId the session
     * @param breakpoints the list of breakpoints
     */
    record BreakpointList(
        String sessionId,
        List<DebugCommand.BreakpointConfig> breakpoints
    ) implements DebugResponse {}

    /**
     * Confirmation that a breakpoint was set.
     *
     * @param sessionId the session
     * @param breakpoint the breakpoint that was set
     */
    record BreakpointSet(
        String sessionId,
        DebugCommand.BreakpointConfig breakpoint
    ) implements DebugResponse {}

    /**
     * Confirmation that a breakpoint was cleared.
     *
     * @param sessionId the session
     * @param breakpointId the ID of the cleared breakpoint
     */
    record BreakpointCleared(
        String sessionId,
        String breakpointId
    ) implements DebugResponse {}

    /**
     * Error response.
     *
     * @param code error code
     * @param message human-readable error message
     * @param sessionId related session if applicable
     */
    record Error(
        String code,
        String message,
        String sessionId
    ) implements DebugResponse {}

    /**
     * List of available session archives.
     *
     * @param archives list of archive summaries
     * @param storageAvailable whether archive storage is configured
     */
    record ArchiveList(List<ArchiveSummary> archives, boolean storageAvailable) implements DebugResponse {}

    /**
     * Confirmation that an archive was imported.
     *
     * @param sessionId the imported session ID
     * @param netName the Petri net name
     * @param eventCount number of events imported
     */
    record ArchiveImported(String sessionId, String netName, long eventCount) implements DebugResponse {}

    /**
     * Summary of an archived session.
     *
     * @param sessionId the session identifier
     * @param key the storage key/path
     * @param sizeBytes file size in bytes
     * @param lastModified when the archive was last modified (ISO-8601)
     */
    record ArchiveSummary(String sessionId, String key, long sizeBytes, String lastModified) {}

    // ======================== Supporting Types ========================

    /**
     * Summary of a debug session for listing.
     *
     * <p>{@code tags}, {@code endTime}, and {@code durationMs} were added in libpetri 1.6.0.
     * Older clients that receive additional fields will ignore them (Jackson tolerance).
     *
     * @param sessionId unique identifier
     * @param netName name of the Petri net
     * @param startTime when the session started (ISO-8601 format)
     * @param active whether the session is still running
     * @param eventCount number of events captured
     * @param tags user-defined session tags (empty map if none)
     * @param endTime when the session ended (ISO-8601), null while active
     * @param durationMs session duration in milliseconds, null while active
     */
    @JsonInclude(JsonInclude.Include.NON_NULL)
    record SessionSummary(
        String sessionId,
        String netName,
        String startTime,
        boolean active,
        long eventCount,
        @JsonInclude(JsonInclude.Include.NON_EMPTY) Map<String, String> tags,
        String endTime,
        Long durationMs
    ) {}

    /**
     * Token information for display and downstream consumption.
     *
     * @param id unique token ID (if available)
     * @param type token value type simple name (display-friendly, LLM-budget-conscious)
     * @param value full {@code toString()} of the token value — stable display field that
     *     the bundled debug UI relies on; always populated unless {@code compact} mode
     *     strips values explicitly
     * @param structured structured JSON representation of the token value when Jackson
     *     can serialize it, else {@code null}. Populated in addition to {@code value} so
     *     LLM-facing MCP tools can surface typed fields (e.g. a {@code ClientMessage}'s
     *     {@code message} field) without parsing the toString string. Omitted from the
     *     wire when null via {@code @JsonInclude(NON_NULL)} at the consumer side
     * @param timestamp when token was created (ISO-8601 format)
     */
    record TokenInfo(
        String id,
        String type,
        String value,
        Object structured,
        String timestamp
    ) {
        /**
         * Backwards-compatible convenience constructor used by callers that only produce the
         * display-form token info (no structured payload). Keeps {@link org.libpetri.debug.DebugProtocolHandler}
         * and other legacy callers compiling unchanged.
         */
        public TokenInfo(String id, String type, String value, String timestamp) {
            this(id, type, value, null, timestamp);
        }
    }

    /**
     * Serializable representation of NetEvent.
     *
     * @param type event type name
     * @param timestamp when the event occurred (ISO-8601 format)
     * @param transitionName transition name (if applicable)
     * @param placeName place name (if applicable)
     * @param details additional event details
     */
    record NetEventInfo(
        String type,
        String timestamp,
        String transitionName,
        String placeName,
        Map<String, Object> details
    ) {}

    // ======================== Net Structure Types ========================

    /**
     * Explicit Petri net structure with authoritative name-to-graphId mappings.
     *
     * <p>This provides the frontend with a reliable mapping between the authoritative
     * place/transition names (used in events and marking) and the sanitized IDs used
     * in the DOT/SVG output. This eliminates fragile SVG ID parsing in the frontend.
     *
     * @param places list of places in the net
     * @param transitions list of transitions in the net
     */
    record NetStructure(
        List<PlaceInfo> places,
        List<TransitionInfo> transitions
    ) {}

    /**
     * Information about a place in the Petri net.
     *
     * <p>The {@code instancePrefix} field is derived from the place's name
     * per {@code spec/11-modular-composition.md} <b>MOD-041</b>: it is the
     * substring before the last {@code "/"}, or {@code null} for flat
     * names. Jackson serializes the {@code null} field as omitted via
     * {@code @JsonInclude(NON_NULL)} so flat-net payloads stay compact and
     * the wire shape pre-MOD-041 is preserved for non-composed nets. Use
     * {@link #instancePrefix()} to retrieve the field as an
     * {@link Optional} when reading from Java.
     *
     * @param name authoritative identifier (matches events/marking keys)
     * @param graphId sanitized ID with "p_" prefix (matches DOT/SVG node IDs)
     * @param tokenType simple name of the token type for this place
     * @param isStart true if this is a start place (no incoming arcs)
     * @param isEnd true if this is an end place (no outgoing arcs)
     * @param isEnvironment true if this is an environment place (external event injection)
     * @param instancePrefixOrNull nullable instance prefix; per <b>MOD-041</b>
     */
    record PlaceInfo(
        String name,
        String graphId,
        String tokenType,
        boolean isStart,
        boolean isEnd,
        boolean isEnvironment,
        @JsonInclude(JsonInclude.Include.NON_NULL)
        @com.fasterxml.jackson.annotation.JsonProperty("instancePrefix")
        String instancePrefixOrNull
    ) {
        /**
         * Backwards-compatible constructor used by call sites that pre-date
         * MOD-041. Sets {@code instancePrefix} to absent (null on the wire).
         */
        public PlaceInfo(String name, String graphId, String tokenType,
                         boolean isStart, boolean isEnd, boolean isEnvironment) {
            this(name, graphId, tokenType, isStart, isEnd, isEnvironment, (String) null);
        }

        /**
         * Convenience constructor that accepts an {@link Optional} for
         * call sites that already work with the optional API shape.
         */
        public PlaceInfo(String name, String graphId, String tokenType,
                         boolean isStart, boolean isEnd, boolean isEnvironment,
                         Optional<String> instancePrefix) {
            this(name, graphId, tokenType, isStart, isEnd, isEnvironment,
                instancePrefix == null ? null : instancePrefix.orElse(null));
        }

        /** Returns the derived instance prefix per [MOD-041] as an {@link Optional}. */
        @com.fasterxml.jackson.annotation.JsonIgnore
        public Optional<String> instancePrefix() {
            return Optional.ofNullable(instancePrefixOrNull);
        }
    }

    /**
     * Information about a transition in the Petri net.
     *
     * @param name authoritative identifier (matches events)
     * @param graphId sanitized ID with "t_" prefix (matches DOT/SVG node IDs)
     * @param instancePrefixOrNull nullable instance prefix; per <b>MOD-041</b>
     */
    record TransitionInfo(
        String name,
        String graphId,
        @JsonInclude(JsonInclude.Include.NON_NULL)
        @com.fasterxml.jackson.annotation.JsonProperty("instancePrefix")
        String instancePrefixOrNull
    ) {
        public TransitionInfo(String name, String graphId) {
            this(name, graphId, (String) null);
        }

        public TransitionInfo(String name, String graphId, Optional<String> instancePrefix) {
            this(name, graphId, instancePrefix == null ? null : instancePrefix.orElse(null));
        }

        /** Returns the derived instance prefix per [MOD-041] as an {@link Optional}. */
        @com.fasterxml.jackson.annotation.JsonIgnore
        public Optional<String> instancePrefix() {
            return Optional.ofNullable(instancePrefixOrNull);
        }
    }

    /**
     * Wire-facing descriptor of one composed subnet instance per
     * {@code spec/11-modular-composition.md} <b>MOD-041</b>.
     *
     * <p>This mirrors {@link org.libpetri.core.SubnetInstance} but carries
     * <b>names</b> instead of object references, so that JSON serialization
     * stays simple (sets of strings rather than recursively-serialized
     * Place/Transition records). {@code parentPrefix} is null on the wire
     * for top-level instances and serialized via {@code @JsonInclude(NON_NULL)}
     * to keep flat-shape payloads compact.
     *
     * @param prefix the instantiation prefix (e.g. {@code "buf1"} or
     *               {@code "outer/inner"} for nested instances)
     * @param defName originating subnet definition name, or {@code null}
     *                when not derivable from name shape alone
     * @param transitionNames full prefixed transition names belonging to this instance
     * @param exposedPlaceNames full prefixed place names belonging to this instance
     * @param parentPrefixOrNull nullable parent-instance prefix; null for top-level
     */
    record SubnetInstanceInfo(
        String prefix,
        @JsonInclude(JsonInclude.Include.NON_NULL) String defName,
        Set<String> transitionNames,
        Set<String> exposedPlaceNames,
        @JsonInclude(JsonInclude.Include.NON_NULL)
        @com.fasterxml.jackson.annotation.JsonProperty("parentPrefix")
        String parentPrefixOrNull
    ) {
        public SubnetInstanceInfo {
            transitionNames = Set.copyOf(transitionNames);
            exposedPlaceNames = Set.copyOf(exposedPlaceNames);
        }

        /**
         * Convenience constructor accepting an {@link Optional} for the
         * {@code parentPrefix} field — call sites that already work with the
         * optional shape can pass it directly.
         */
        public SubnetInstanceInfo(String prefix, String defName,
                                  Set<String> transitionNames, Set<String> exposedPlaceNames,
                                  Optional<String> parentPrefix) {
            this(prefix, defName, transitionNames, exposedPlaceNames,
                parentPrefix == null ? null : parentPrefix.orElse(null));
        }

        /** Returns the parent prefix per [MOD-041] as an {@link Optional}. */
        @com.fasterxml.jackson.annotation.JsonIgnore
        public Optional<String> parentPrefix() {
            return Optional.ofNullable(parentPrefixOrNull);
        }
    }
}

File: rust/libpetri-debug/src/debug_response.rs
//! Responses sent from server to debug UI client via WebSocket.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::debug_command::{BreakpointConfig, EventFilter};

/// Summary of a debug session.
///
/// `tags`, `end_time`, and `duration_ms` were added in libpetri 1.6.0. Older clients that
/// do not know these fields ignore them (serde tolerance).
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummary {
    pub session_id: String,
    pub net_name: String,
    pub start_time: String,
    pub active: bool,
    pub event_count: usize,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub tags: HashMap<String, String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
}

/// Serializable token information.
///
/// `structured` (libpetri 1.8.0+) carries a typed JSON projection of the token
/// value produced by [`TokenProjectorRegistry`](crate::TokenProjectorRegistry).
/// Populated alongside `value` (not instead of) so the bundled debug UI keeps
/// rendering unchanged while LLM-facing consumers get typed fields. Omitted
/// from the wire when absent.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenInfo {
    pub id: Option<String>,
    #[serde(rename = "type")]
    pub token_type: String,
    pub value: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub structured: Option<serde_json::Value>,
    pub timestamp: Option<String>,
}

/// Serializable event information.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NetEventInfo {
    #[serde(rename = "type")]
    pub event_type: String,
    pub timestamp: String,
    pub transition_name: Option<String>,
    pub place_name: Option<String>,
    pub details: HashMap<String, serde_json::Value>,
}

/// Information about a place in the net structure.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaceInfo {
    pub name: String,
    pub graph_id: String,
    pub token_type: String,
    pub is_start: bool,
    pub is_end: bool,
    pub is_environment: bool,
}

/// Information about a transition in the net structure.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransitionInfo {
    pub name: String,
    pub graph_id: String,
}

/// Structure of a Petri net for the debug UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NetStructure {
    pub places: Vec<PlaceInfo>,
    pub transitions: Vec<TransitionInfo>,
}

/// Summary of an archived session.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArchiveSummary {
    pub session_id: String,
    pub key: String,
    pub size_bytes: u64,
    pub last_modified: String,
}

/// Responses from server to debug UI client.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum DebugResponse {
    SessionList {
        sessions: Vec<SessionSummary>,
    },
    Subscribed {
        session_id: String,
        net_name: String,
        dot_diagram: String,
        structure: NetStructure,
        current_marking: HashMap<String, Vec<TokenInfo>>,
        enabled_transitions: Vec<String>,
        in_flight_transitions: Vec<String>,
        event_count: usize,
        mode: String,
    },
    Unsubscribed {
        session_id: String,
    },
    Event {
        session_id: String,
        index: usize,
        event: NetEventInfo,
    },
    EventBatch {
        session_id: String,
        start_index: usize,
        events: Vec<NetEventInfo>,
        has_more: bool,
    },
    MarkingSnapshot {
        session_id: String,
        marking: HashMap<String, Vec<TokenInfo>>,
        enabled_transitions: Vec<String>,
        in_flight_transitions: Vec<String>,
    },
    PlaybackStateChanged {
        session_id: String,
        paused: bool,
        speed: f64,
        current_index: usize,
    },
    FilterApplied {
        session_id: String,
        filter: EventFilter,
    },
    BreakpointHit {
        session_id: String,
        breakpoint_id: String,
        event: NetEventInfo,
        event_index: usize,
    },
    BreakpointList {
        session_id: String,
        breakpoints: Vec<BreakpointConfig>,
    },
    BreakpointSet {
        session_id: String,
        breakpoint: BreakpointConfig,
    },
    BreakpointCleared {
        session_id: String,
        breakpoint_id: String,
    },
    Error {
        code: String,
        message: String,
        session_id: Option<String>,
    },
    ArchiveList {
        archives: Vec<ArchiveSummary>,
        storage_available: bool,
    },
    ArchiveImported {
        session_id: String,
        net_name: String,
        event_count: usize,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serde_round_trip_session_list() {
        let resp = DebugResponse::SessionList {
            sessions: vec![SessionSummary {
                session_id: "s1".into(),
                net_name: "test".into(),
                start_time: "2025-01-01T00:00:00Z".into(),
                active: true,
                event_count: 42,
                ..Default::default()
            }],
        };
        let json = serde_json::to_string(&resp).unwrap();
        assert!(json.contains("\"type\":\"sessionList\""));
        // New 1.6.0 fields are omitted when unset (skip_serializing_if)
        assert!(!json.contains("\"tags\""));
        assert!(!json.contains("\"endTime\""));
        assert!(!json.contains("\"durationMs\""));
        let back: DebugResponse = serde_json::from_str(&json).unwrap();
        match back {
            DebugResponse::SessionList { sessions } => {
                assert_eq!(sessions.len(), 1);
                assert_eq!(sessions[0].session_id, "s1");
                assert!(sessions[0].tags.is_empty());
                assert!(sessions[0].end_time.is_none());
                assert!(sessions[0].duration_ms.is_none());
            }
            _ => panic!("wrong variant"),
        }
    }

    #[test]
    fn serde_round_trip_session_summary_with_1_6_0_fields() {
        let mut tags = HashMap::new();
        tags.insert("channel".to_string(), "voice".to_string());
        tags.insert("env".to_string(), "staging".to_string());

        let summary = SessionSummary {
            session_id: "s1".into(),
            net_name: "test".into(),
            start_time: "1000".into(),
            active: false,
            event_count: 42,
            tags,
            end_time: Some("2500".into()),
            duration_ms: Some(1500),
        };
        let json = serde_json::to_string(&summary).unwrap();

        assert!(json.contains("\"tags\""));
        assert!(json.contains("\"channel\":\"voice\""));
        assert!(json.contains("\"endTime\":\"2500\""));
        assert!(json.contains("\"durationMs\":1500"));

        let back: SessionSummary = serde_json::from_str(&json).unwrap();
        assert_eq!(back.session_id, "s1");
        assert_eq!(back.end_time, Some("2500".into()));
        assert_eq!(back.duration_ms, Some(1500));
        assert_eq!(back.tags.get("channel"), Some(&"voice".to_string()));
        assert_eq!(back.tags.get("env"), Some(&"staging".to_string()));
    }

    #[test]
    fn serde_deserialize_session_summary_tolerates_missing_1_6_0_fields() {
        // An old-format payload (pre-1.6.0) must deserialize cleanly.
        let json = r#"{
            "sessionId": "s1",
            "netName": "test",
            "startTime": "1000",
            "active": true,
            "eventCount": 0
        }"#;
        let summary: SessionSummary = serde_json::from_str(json).unwrap();
        assert_eq!(summary.session_id, "s1");
        assert!(summary.tags.is_empty());
        assert!(summary.end_time.is_none());
        assert!(summary.duration_ms.is_none());
    }

    #[test]
    fn serde_round_trip_subscribed() {
        let resp = DebugResponse::Subscribed {
            session_id: "s1".into(),
            net_name: "test".into(),
            dot_diagram: "digraph {}".into(),
            structure: NetStructure {
                places: vec![PlaceInfo {
                    name: "p1".into(),
                    graph_id: "p_p1".into(),
                    token_type: "i32".into(),
                    is_start: true,
                    is_end: false,
                    is_environment: false,
                }],
                transitions: vec![TransitionInfo {
                    name: "t1".into(),
                    graph_id: "t_t1".into(),
                }],
            },
            current_marking: HashMap::new(),
            enabled_transitions: vec!["t1".into()],
            in_flight_transitions: vec![],
            event_count: 5,
            mode: "live".into(),
        };
        let json = serde_json::to_string(&resp).unwrap();
        assert!(json.contains("\"type\":\"subscribed\""));
        let _back: DebugResponse = serde_json::from_str(&json).unwrap();
    }

    #[test]
    fn serde_round_trip_error() {
        let resp = DebugResponse::Error {
            code: "NOT_FOUND".into(),
            message: "Session not found".into(),
            session_id: Some("s1".into()),
        };
        let json = serde_json::to_string(&resp).unwrap();
        let back: DebugResponse = serde_json::from_str(&json).unwrap();
        match back {
            DebugResponse::Error {
                code,
                message,
                session_id,
            } => {
                assert_eq!(code, "NOT_FOUND");
                assert_eq!(message, "Session not found");
                assert_eq!(session_id, Some("s1".into()));
            }
            _ => panic!("wrong variant"),
        }
    }

    #[test]
    fn serde_response_inline_fields_use_camelcase() {
        // Cross-language interop: inline enum-variant fields must serialize as camelCase
        // so the TypeScript debug-ui (and Java client) can parse them.
        let resp = DebugResponse::Subscribed {
            session_id: "s1".into(),
            net_name: "test".into(),
            dot_diagram: "digraph {}".into(),
            structure: NetStructure {
                places: vec![],
                transitions: vec![],
            },
            current_marking: HashMap::new(),
            enabled_transitions: vec!["t1".into()],
            in_flight_transitions: vec![],
            event_count: 5,
            mode: "live".into(),
        };
        let json = serde_json::to_string(&resp).unwrap();
        assert!(json.contains("\"sessionId\":\"s1\""));
        assert!(json.contains("\"netName\":\"test\""));
        assert!(json.contains("\"dotDiagram\""));
        assert!(json.contains("\"currentMarking\""));
        assert!(json.contains("\"enabledTransitions\""));
        assert!(json.contains("\"inFlightTransitions\""));
        assert!(json.contains("\"eventCount\":5"));
        assert!(!json.contains("\"session_id\""));
        assert!(!json.contains("\"net_name\""));
        assert!(!json.contains("\"event_count\""));
    }

    #[test]
    fn serde_event_batch_camelcase_interop() {
        // Cross-language interop: a Java/TS-shaped EventBatch payload must deserialize.
        let json = r#"{"type":"eventBatch","sessionId":"s1","startIndex":0,"events":[],"hasMore":true}"#;
        let resp: DebugResponse = serde_json::from_str(json).unwrap();
        match resp {
            DebugResponse::EventBatch {
                session_id,
                start_index,
                has_more,
                ..
            } => {
                assert_eq!(session_id, "s1");
                assert_eq!(start_index, 0);
                assert!(has_more);
            }
            _ => panic!("wrong variant"),
        }
    }

    #[test]
    fn serde_all_response_variants() {
        let responses: Vec<DebugResponse> = vec![
            DebugResponse::SessionList { sessions: vec![] },
            DebugResponse::Unsubscribed {
                session_id: "s1".into(),
            },
            DebugResponse::Event {
                session_id: "s1".into(),
                index: 0,
                event: NetEventInfo {
                    event_type: "TransitionStarted".into(),
                    timestamp: "2025-01-01T00:00:00Z".into(),
                    transition_name: Some("t1".into()),
                    place_name: None,
                    details: HashMap::new(),
                },
            },
            DebugResponse::EventBatch {
                session_id: "s1".into(),
                start_index: 0,
                events: vec![],
                has_more: false,
            },
            DebugResponse::MarkingSnapshot {
                session_id: "s1".into(),
                marking: HashMap::new(),
                enabled_transitions: vec![],
                in_flight_transitions: vec![],
            },
            DebugResponse::PlaybackStateChanged {
                session_id: "s1".into(),
                paused: true,
                speed: 1.0,
                current_index: 0,
            },
            DebugResponse::FilterApplied {
                session_id: "s1".into(),
                filter: EventFilter::all(),
            },
            DebugResponse::BreakpointList {
                session_id: "s1".into(),
                breakpoints: vec![],
            },
            DebugResponse::BreakpointCleared {
                session_id: "s1".into(),
                breakpoint_id: "bp1".into(),
            },
            DebugResponse::Error {
                code: "ERR".into(),
                message: "msg".into(),
                session_id: None,
            },
            DebugResponse::ArchiveList {
                archives: vec![],
                storage_available: false,
            },
            DebugResponse::ArchiveImported {
                session_id: "s1".into(),
                net_name: "test".into(),
                event_count: 10,
            },
        ];
        for resp in responses {
            let json = serde_json::to_string(&resp).unwrap();
            let _back: DebugResponse = serde_json::from_str(&json).unwrap();
        }
    }
}
